interactionlib.php

```diff
@@ -0,0 +1,193 @@
+        } else {
+            return array(
+                array('action' => self::ACTION_NONOTIFY,
+                    'alt' => get_string('nonotify', 'lifecyclestep_notifystudents'),
+                ),
+            );
+        }
+    }
+
+    /**
+     * Returns the status message for the given process.
+     * @param process $process process the status message is requested for
+     * @return string status message
+     * @throws \coding_exception
+     */
+    public function get_status_message($process) {
+        $step = step_manager::get_step_instance_by_workflow_index($process->workflowid, $process->stepindex);
+        $settings = settings_manager::get_settings($step->id, settings_type::STEP);
+        if ($settings['option'] == 1) {
+            return get_string('status_message_nonotify', 'lifecyclestep_notifystudents');
+        } else {
+            return get_string('status_message_notify', 'lifecyclestep_notifystudents');
+        }
+    }
+
+    /**
+     * Called when a user triggered an action for a process instance.
+     * @param process $process instance of the process the action was triggered upon.
+     * @param step_subplugin $step instance of the step the process is currently in.
+     * @param string $action action string
+     * @return step_interactive_response defines if the step still wants to process this course
+     *      - proceed: the step has finished and respective controller class can take over.
+     *      - stillprocessing: the step still wants to process the course and is responsible for rendering the site.
+     *      - noaction: the action is not defined for the step.
+     *      - rollback: the step has finished and respective controller class should rollback the process.
+     */
+    public function handle_interaction($process, $step, $action = 'default') {
+        global $DB, $PAGE;
+        $PAGE->set_context(\context_system::instance());
+        $settings = settings_manager::get_settings($step->id, settings_type::STEP);
+        $type = 'student';
+        $typeid = 1;
+        if ($action == self::ACTION_NOTIFY) {
+            $settings[$type . '_content'] = format_text($settings[$type . '_content'], FORMAT_HTML);
+            $userstobeinformed = $DB->get_records('lifecyclestep_notifystudents',
+                array('courseid' => $process->courseid, 'instanceid' => $step->id, 'emailtype' => $typeid), '', 'distinct touser');
+            foreach ($userstobeinformed as $userrecord) {
+                $user = \core_user::get_user($userrecord->touser);
+                $transaction = $DB->start_delegated_transaction();
+                $mailentries = $DB->get_records('lifecyclestep_notifystudents',
+                    array('instanceid' => $step->id, 'courseid' => $process->courseid,
+                        'touser' => $user->id, 'emailtype' => $typeid));
+
+                $parsedsettings = (new notifystudents)->replace_placeholders($settings, $user, $step->id, $mailentries);
```

Comment:
If i see correctly, this (Lines 124-143) is also more or less the same code as the loop in the send_email function, could you refactor that code into one function?

Reply:
> I was surprised that php allows this, but I think you should write `(new notifystudents())` (or consider making replace_placeholders (, and parse_course_row_data) static) :)

made the functions static.



> If i see correctly, this (Lines 124-143) is also more or less the same code as the loop in the send_email function, could you refactor that code into one function?

I was thinking about it, but was hesitating because in lib.php all remaining db entries are considered wheras in the interactionlib.php only the action course is considered. I will see if I find I solution. I might refactor parts of the function, but not the whole thing.

Reply:
Ah, I see. I did not see that difference. But then, the inner loop is unnecessary, because there could only be one $mailentry, right? If that is correct I would say it would be best either to edit this function, so that it doesn't have an inner loop, or refactor it (with parameter $courseid = null, the db parameter array defined as an variable beforehand and if $courseid is specified, it is added to the parameter array)

Reply:
I don't really understand, which loop you are referring to because the foreach loops function here is to iterate through the course users and not through the mail entries. The difference between the variable $mailentries here versus in lib.php is that there will always be just one entry for each user. The replace_placeholder function will have an unnecessary loop in this case, but i don't think it is necessary to edit it for this specific case.

Reply:
Sorry, my comment does not really make sense. I meant that the db query inside the loop is unnecessary (exactly because there is only one entry per user). So just selecting all mail entries before the loop, and deleting all after the loop would be enough, correct?

So I would say you can choose if you do that, but if not, I would suggest refactoring the code, because then the two code sections are very similar

Reply:
I'm not sure if completely understand what you are saying but it does not really matter because I already refactored the code and added another optional variable §courseid to the send_email() function.